### PR TITLE
CTCP-3844-1: MoreInformation Controller

### PR DESCRIPTION
**Pull request by justega247**

### CODE REVIEW

File: app/controllers/IndexController.scala [hunk 1]
```diff
@@ -42,25 +42,33 @@ class IndexController @Inject() (
...
+                .map {
+                  _ =>
+                    if (departureData.data.isDataComplete) {
+                      Redirect(Call("GET", "#"))
+                    } else {
+                      Redirect(controllers.routes.MoreInformationController.onPageLoad(departureId))
+                    }
+                }
+          }
+        case None => ??? //technical difficulties
```
Comment (MatthewAlexanderBeck) [inline app/controllers/IndexController.scala:70]: Can we add this in quickly? Dont want to kick the can further down the road: 
https://github.com/hmrc/manage-transit-movements-departure-frontend/blob/main/app/controllers/ErrorController.scala

File: app/controllers/IndexController.scala [hunk 2]
```diff
@@ -42,25 +42,33 @@ class IndexController @Inject() (
...
                 )
-              )
-            )
-            .map(
-              _ => Redirect(controllers.routes.MoreInformationController.onPageLoad(departureId))
-            )
+                .map {
+                  _ =>
+                    if (departureData.data.isDataComplete) {
+                      Redirect(Call("GET", "#"))
```
Comment (MatthewAlexanderBeck) [inline app/controllers/IndexController.scala:64]: I think Ken has merged this page. If you pull in main you can get this to redirect to the correct place.

File: app/controllers/IndexController.scala [hunk 3]
```diff
@@ -42,25 +42,33 @@ class IndexController @Inject() (
...
-              request.userAnswers.getOrElse(
-                UserAnswers(
-                  departureId,
-                  request.eoriNumber,
-                  lrn.value,
-                  JsObject.empty,
-                  timeMachine.now()
+      service.getDepartureData(departureId).flatMap {
+        case Some(departureData) =>
+          service.getLRN(departureId) flatMap {
```
Comment (MatthewAlexanderBeck) [inline app/controllers/IndexController.scala:47]: We can retrieve the LRN from the MessageMetaData so we can remove this call and add it to the reads for MessageMetaData:
https://developer.service.hmrc.gov.uk/api-documentation/docs/api/service/common-transit-convention-traders/2.0/oas/page#tag/Departures/operation/getCachedDepartureForDepartureID

File: app/models/DepartureSubmission.scala [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+
+package models
+
+import play.api.libs.json.Reads
+
+trait DepartureSubmission
+
+//object DepartureSubmission {
+//  implicit val reads: Reads[]
+//}
```
Comment (MatthewAlexanderBeck) [inline app/models/DepartureSubmission.scala:25]: Is this needed?

File: app/models/messages/Data.scala [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+
+case class Data(data: MessageData)
+
+object Data {
+
+  def reads(messageType: DepartureMessageType): Reads[Data] =
+    messageType match {
+      case DepartureNotification => (__ \ "body" \ "n1:CC015C").read[MessageData].map(Data.apply)
+      case AmendmentSubmitted    => (__ \ "body" \ "n1:CC013C").read[MessageData].map(Data.apply)
+    }
```
Comment (MatthewAlexanderBeck) [inline app/models/messages/Data.scala:31]: A refactor idea would be to have a 'toString' like method on DepartureMessageType and append the path e.g

(__ \ "body" \  s"n1:${messageType.toString}").read[MessageData].map(Data.apply)

Also we will get a matcher error here. Maybe have a catch all that will fail the reads if we its a different matcher error

File: app/connectors/DepartureMovementConnector.scala [hunk 6]
```diff
@@ -29,6 +31,12 @@ class DepartureMovementConnector @Inject() (config: FrontendAppConfig, http: Htt
 
   private def headers(implicit hc: HeaderCarrier): HeaderCarrier = hc.withExtraHeaders(("Accept", "application/vnd.hmrc.2.0+json"))
 
+  def getData(location: String, messageType: DepartureMessageType)(implicit hc: HeaderCarrier): Future[Data] = {
+    implicit val dataReads: Reads[Data] = Data.reads(messageType)
+    val url                             = s"${config.commonTransitConventionTradersUrl}$location"
+    http.GET[Data](url)(implicitly, headers, ec)
+  }
+
```
Comment (MatthewAlexanderBeck) [inline app/connectors/DepartureMovementConnector.scala:39]: Can we add a unit test for this please? :)

**Review by jackheslop96 — APPROVED** [review]

**Review by jackheslop96 — APPROVED** [review]

**Pull request merged**